narwhal/primary/src/synchronizer.rs

```diff
@@ -393,90 +411,133 @@ impl Synchronizer {
-            for (name, _, network_key) in inner_senders
-                .committee
-                .others_primaries_by_id(inner_senders.authority_id)
-                .into_iter()
-            {
-                senders.spawn(Self::push_certificates(
-                    network.clone(),
-                    name,
-                    network_key,
-                    tx_own_certificate_broadcast.subscribe(),
-                ));
-            }
-            if let Some(cert) = highest_created_certificate {
-                // Error can be ignored.
-                if tx_own_certificate_broadcast.send(cert).is_err() {
-                    error!("Failed to populate initial certificate to send to peers!");
+                let mut senders = inner_senders.certificate_senders.lock();
+                for (name, _, network_key) in inner_senders
+                    .committee
+                    .others_primaries_by_id(inner_senders.authority_id)
+                    .into_iter()
+                {
+                    senders.spawn(Self::push_certificates(
+                        network.clone(),
+                        name,
+                        network_key,
+                        tx_own_certificate_broadcast.subscribe(),
+                    ));
                 }
-            }
-        });
+                if let Some(cert) = highest_created_certificate {
+                    // Error can be ignored.
+                    if tx_own_certificate_broadcast.send(cert).is_err() {
+                        error!("Failed to populate initial certificate to send to peers!");
+                    }
+                }
+            },
+            "SynchronizerCertificateBroadcasters"
+        );
+
+        // Start a task to async download batches if needed
+        let weak_inner = Arc::downgrade(&inner);
+        spawn_logged_monitored_task!(
+            async move {
+                let mut batch_tasks: JoinSet<DagResult<()>> = JoinSet::new();
+
+                loop {
+                    tokio::select! {
+                        Some((header, max_age, certified)) = rx_batch_tasks.recv() => {
+                            let Some(inner) = weak_inner.upgrade() else {
+                                debug!("Synchronizer is shutting down.");
+                                return;
+                            };
+
+                            batch_tasks.spawn(async move {
+                                Synchronizer::sync_batches_internal(inner.clone(), &header, max_age, certified).await
+                            });
+                        },
+                        Some(result) = batch_tasks.join_next() => {
+                            if let Err(err) = result  {
+                                error!("Error when synchronizing batches: {err:?}")
+                            }
+                        },
+                        else => {
+                            break;
+                        }
```

Comment:
do you need the else?

Reply:
Yep, the `else` is needed when eventually all the branches become disabled during shutdown since the channel closes and the joinset is consumed (we should always having at least one active branch), otherwise we'll panic. Now, based on Mingwei's comment bellow it's probably good idea to handle the `None` case from the returned channel, and then I believe `else` here will become redundant.